lib/dynamoid/transaction/executor.rb

```diff
@@ -0,0 +1,78 @@
+
+require 'dynamoid/transaction/action'
+require 'dynamoid/transaction/create'
+require 'dynamoid/transaction/delete'
+require 'dynamoid/transaction/update'
+require 'dynamoid/transaction/upsert'
+
+module Dynamoid
+  module Transaction
+    class Executor
+      def initialize(actions = [], options = {}, &block)
+        @skip_callbacks = !!options[:skip_callbacks]
+        @skip_validation = !!options[:skip_validation]
+        @actions = actions
+        instance_eval &block if block
+      end
+
+      def call
+        validate! unless @skip_validation
+        call_with_callbacks(@actions)
+      end
+
+      # throws Dynamoid::Errors::DocumentNotValid if model is present and not valid
+      # use skip_validation: true to skip
+      def validate!
+        @actions.each do |action|
+          next if !action.is_a?(Action) || !(model = action.model) || action.skip_validation?
+          raise Dynamoid::Errors::DocumentNotValid.new(model) unless model.valid?
+        end
+      end
+
+      # recursively call save, update, create, and validate callbacks
+      # only 'before' and first half of 'around' filters happen before transaction
+      def call_with_callbacks(actions = [])
+        return call_without_callbacks if actions.empty? # no models left so recursion ends
+        actions = actions.clone
+        action = actions.shift # remove first item
+        if action.respond_to?(:run_callbacks)
+          action.run_callbacks do
+            call_with_callbacks(actions) # recursive call with remaining items
+          end
+        else
+          call_with_callbacks(actions)
+        end
+      end
+
+      def call_without_callbacks
+        puts "call_without_callbacks actions #{@actions}"
+        Dynamoid.adapter.transact_write_items(@actions)
+      end
+
+      def save(model, options = {})
+        model.new_record? ? create(model, {}, options) : update(model, {}, options)
+      end
+
+      def create(model_or_model_class, attributes = {}, options = {})
+        @actions << Dynamoid::Transaction::Create.new(model_or_model_class, attributes, options)
+      end
+
+      def upsert(model_or_model_class, attributes = {}, options = {})
+        @actions << Dynamoid::Transaction::Upsert.new(model_or_model_class, attributes, options)
+      end
+
+      def update(model_or_model_class, attributes = {}, options = {})
+        @actions << Dynamoid::Transaction::Update.new(model_or_model_class, attributes, options)
+      end
+
+      def delete(model_or_model_class, attributes = {}, options = {})
+        @actions << Dynamoid::Transaction::Delete.new(model_or_model_class, attributes, options)
+      end
```

Comment:
I was able to make the instance and class invocations very similar since the class invocation will always create or find an instance first. However, none of these accept blocks at the moment so that could change.